UI: Add ability to copy scenes to other scene collections

RytoEX opened the pull request:
**Note:  This PR is not ready to be merged.**  I'm submitting it to get some eyes on it for review, testing, and cleanup.

This commit adds the ability for users to copy an entire scene and all of its child sources to another scene collection. This is for parity with OBS Classic.

I've been staring at this code and adjusting it on and off for a few months.  The functionality works, and I've compiled and tested it on Windows 10 64-bit.  At this point I've been looking at it so long that my eyes are just glossing over it now, so I'm in need of some fresh perspective.  I could use some feedback on everything here - project code style compliance, efficiency, lang strings, function/variable names, expected/actual functionality, user experience, etc.  I'm aware that there are still a lot of comments in the code.  Most of those I left in place intentionally for the purpose of this review.

To preemptively answer the questions, "Why so much new code? Why not use existing functions?"  Most of the new code in this PR was written because existing functions for copying scenes/sources seemed to rely on or affect the current UI state.  I needed to load and save data without those caveats, and I didn't want to alter any existing functions.

Any feedback is appreciated, though constructive feedback would _really_ be appreciated.  I will try to answer any questions as best as I can.

SuslikV commented:
I think, "Duplicate to..." ("..." at the end because of new dialog window) is better name for the option. At least, OBS Studio already has "Duplicate" option by right mouse click over the Scene.

RytoEX commented:
I've fixed the merge conflicts with master, and I've adjusted this patch to account for recent changes due to PR #917.  These fixes are minor and will be pushed as an update to the original commit.

After further testing of my own, I discovered that the original patch does not account for sources of type "scene" that are within a scene being copied.  I have fixed this, and will push a new commit shortly.  **_For now_**, I'll keep it as a separate commit only to make it obvious what I had to add to fix that issue.  Again, keeping the separate commit is only temporary - the commits will be squashed later.  I'm not terribly pleased with the additional fix, as I think it's a bit ugly.  Hopefully, someone will have some constructive feedback on how to improve it, or I'll have some new insight later.

I've also considered that I can probably remove all or most of the code in [OBSBasic::AddSceneToSceneCollection](https://github.com/jp9000/obs-studio/pull/864/files#diff-a0eec92bde9402a347f40d7593c0cdc9R2027) related to `saveData` and `GenerateSaveData`.  Instead, I can just use `obs_data_set_array` to save the modified `destSources` source array back into `loadData` and save that back to the target Scene Collection JSON file.  When I originally wrote this patch, I think that I thought I might need some of the scene collection data later, so I kept the loaded data and the data intended for saving separate.  Thinking on it now, I don't see why I would need the data separate, and it just adds to the size (about 70 lines) and complexity (having to account for new/modified settings, like PR #917) of the patch.  I'll test this idea later.  Right now, I just wanted to address the immediately apparent issues with this PR that needed resolved.

---

@SuslikV
I don't like the "Duplicate to..." wording for a two reasons:
1. It doesn't immediately indicate what kind of destination to which the user is copying the scene.
2. The word "duplicate" is a more complex word than "copy" (according to language complexity analysis tools).

reboot commented:
I would recommand to exclude the current scene collection from the dropdown for selecting the destination scene collection. It makes no sense to copy a scene to the same scene collection. Also you could probably combine the scene collection and scene name dialog into a single dialog.

RytoEX commented:
@reboot 
I implemented your first suggestion (exclude the current scene collection from the dropdown for selecting the destination scene collection), but it seems there's a memory leak that appears in the log output when I call `OBSStudioAPI::obs_frontend_get_current_scene_collection` to get the current scene collection name.  If I call `config_get_string(App()->GlobalConfig(), "Basic", "SceneCollection")` directly, I don't see the memory leak.  I'm not sure what's causing that, and it's outside the scope of this PR, but @jp9000 might be able to sort that out.  I may hold off on pushing this feature implementation until I have gotten some guidance, or at least until I've slept on it.

As for your second suggestion (combine the scene collection and scene name dialog into a single dialog), I'll look into it.

---

As for [my earlier comment](https://github.com/jp9000/obs-studio/pull/864#issuecomment-315592667), I did test removing the code related to `saveData` and `GenerateSaveData`, and everything still works as expected.  Unless someone can think of a very good reason to keep those parts in the patch, I'll probably continue in that direction.

reboot commented:
`obs_frontend_get_current_scene_collection()` creates a copy of the scene collection name, you probably just have to free it, when you don't need it anymore.

RytoEX commented:
@reboot 
Ah, you're right.  I've fixed that issue in my local copy.  The memory leak no longer appears.

I'm still going over the idea of combining the two dialog boxes into one.  Originally, I'd hoped to keep the SceneCollectionDialog (the first dialog box you see to select the destination Scene Collection) generic enough that it could possibly be reused for other purposes.  I didn't want to make something so specific to my needs that it could _only_ be used for this purpose.  However, if the UX factor outweighs the code-reuse factor here (it very well might), then I'll abandon that thought and see what I can do about simplifying that process.

SuslikV commented:
With two dialog, at least on my PC, I see that something changes, fully redrawn (maybe for faster PC it is not quite visible?) and this rises my attention to the new dialog. If you reuse dialog, then only size of the dialog will update (nobody reads the text labels) - less attention to the changes from the user side. So, leave it as is.

Any chance to implement source auto-renaming (as for copy/paste)?

reboot commented:
You could probably still make the dropdown a reuseable component.

But overall, after thinking about this issue for some time, i think maybe the context menu of a scene is the wrong place to put this function at all. I think it may be better to reverse the whole process, instead of copying a scene from the current scene collection to another scene collection, it may be better to import a scene from another scene collection into the current scene collection. You could then put the function into the Scene Collection menu, maybe like

```
  Scene Collection
   `- ...
   `- Import
       `- Whole Scene Collection
       `- Scene from other Scene Collection
   `- ...
```

The first submenu item is the old import, the second one would be the new function. That would bring up a dialog that lets you select the source collection (maybe even allow selecting an exising scene collection or from an exported scene collection file) and one or multiple scenes to import.

I don't know how much work that would be, or if it is even possible. Just an idea to keep a rarely used function out of the context menu. I think it is more likely that I need a scene from a different scene collection in the current scene collection. With the current appoach, i would have to switch to the other scene collection, copy the scene and then switch back. With an import into the current scene collection, i can just get the scene without switching the collection at all. I could even get multiple scenes from different collections easily.

Allowing an import of multiple scene would also make it easier if 2 scenes share a source, for example a background image. With the current approach you have to do 2 copy operations, but after you completed the first copy, all sources with their names exist in the target scene collection. Copying another scene that shared sources with the first scene will then create conflicts.

SuslikV commented:
Nope. Import is when you can save data to disk. We have "duplicate" option that makes clones of sources automatically (paste as ref.), and you planning to add another option that has strange meaning? I think it is bad idea to make "Import" that can open only OBS own internal data and nothing else.

RytoEX commented:
@reboot
From what I can recall, the workflow you suggest would be more difficult to implement because the scenes are not loaded, so I would not have ready access to initialized scene/source objects to use for things like `obs_scene_enum_items`.  I'm not sure if you can create a source/scene object from JSON data if that source isn't accessible in the currently loaded scene collection.  I would guess that it is possible, since that's probably how OBS would load a scene collection for display.  If you can't, I would have to manually iterate through the scene data to recursively find all of its child sources.  I'll look into that possibility, but I'm not convinced that importing data from unloaded scene collections is the best way to go.  Writing finished JSON data seemed to be the easier solution when I originally looked at this.

That challenge aside, I had originally modeled this after the workflow in OBS Classic, which uses this as a context menu item.  Though, I wasn't able to easily generate a new submenu for the context menu item, and I thought the submenu would get ridiculous for large sets of scene collections, which is why I used dialog boxes instead.

SuslikV commented:
`Basic.Main.CopySceneToSceneCollection="Copy To Scene Collection"`

I think this one is better:
`Basic.Main.CopySceneToSceneCollection="Copy to Scene Collection"`

Look at: Order>Move **t**o Top

@jp9000 Whole project (OBS Studio) don't uses `...` at the end of the menu item name - but the action can be canceled, new dialog window created etc. Is this some kind of styling?

RytoEX commented:
@SuslikV 

`Basic.Main.CopySceneToSceneCollection="Copy to Scene Collection"`

Sounds good to me.  I've made that change locally.

RytoEX commented:
I haven't addressed all issues brought up in this thread, but I've rebased and pushed some additional fixes to try to keep this from falling behind.

ONSBalder commented:
Can one of the admins verify this patch?

RytoEX commented:
I've rebased my branch, squashed the commits, and pushed a single commit to keep this PR updated.

Here's the previous commit history to show what issues I've addressed:
> Fixup: Allow recursively adding sources if source is a scene
Fixup: Remove code related to GenerateSaveData to simplify things
Fixup: Don't allow user to copy scene to current scene collection
Fixup: Locale adjustments

jp9000 commented:
Closing due to extended inactivity.  Feel free to reopen later on.

RytoEX commented:
To follow-up on the discussion that @jp9000 and I had in Discord, since there is [a Fider issue](https://ideas.obsproject.com/ideas/20/add-the-ability-to-copy-and-paste-duplicate-sources-and-or-scenes-across-scene-collections) with a fair number of votes, I'm currently planning to dust this off next month, rework the workflow, and reopen this at that time.
